common/flatpak-dir.c

```diff
@@ -14628,3 +14628,345 @@ flatpak_dir_delete_mirror_refs (FlatpakDir    *self,
 
   return TRUE;
 }
+
+static gboolean
+maybe_get_metakey_and_origin (FlatpakDir  *dir,
+                              const char  *ref,
+                              GHashTable  *metadata_injection,
+                              GKeyFile   **out_metakey,
+                              char       **out_origin)
+{
+  g_autoptr(GKeyFile) metakey = NULL;
+  g_autofree char *origin = NULL;
+
+  origin = flatpak_dir_get_origin (dir, ref, NULL, NULL);
+  if (origin == NULL)
+    return FALSE;
+
+  if (metadata_injection != NULL)
+    metakey = g_hash_table_lookup (metadata_injection, ref);
+  if (metakey != NULL)
+    g_key_file_ref (metakey);
+  else
+    {
+      g_autoptr(FlatpakDeploy) deploy = flatpak_dir_load_deployed (dir, ref, NULL, NULL, NULL);
+      if (deploy == NULL)
+        return FALSE;
+      metakey = flatpak_deploy_get_metadata (deploy);
+    }
+
+  if (out_metakey)
+    *out_metakey = g_steal_pointer (&metakey);
+  if (out_origin)
+    *out_origin = g_steal_pointer (&origin);
+
+  return TRUE;
+}
+
+static void
+find_used_related_refs (FlatpakDir *dir,
+                        FlatpakDir *system_dir, /* nullable */
```

Comment:
I think this naming and use of dir and system_dir is super confusing. Whenever super_dir is set we use it instead of dir with an if, except in one place.

I think always passing the directory we're looking for related refs in as "dir", and then have a "user_dir"  argument which is non-NULL in the case the dependency came from the user dir.